sql: check session_user can become user on deserialize_session

otan opened the pull request:
Release note (bug fix): `crdb_internal.deserialize_session` now checks
the session_user has the privilege to SET ROLE to the current_user before
changing the session settings.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75575)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/testdata/session_migration/errors:
```diff
@@ -67,3 +67,18 @@ SELECT crdb_internal.deserialize_session(
+SELECT crdb_internal.deserialize_session(
+  decode(
+    '0a510a0964656661756c74646212102420636f636b726f6163682064656d6f1a04726f6f742204100222002802380842035554434a0524757365724a067075626c69635a0060808080207a0088010190018050124910904e3002380840026001680170017801880101d80101e00101f00101f80101900201b002808001c80201f202087465737475736572a00301a9030000000000408f40d00301e00301',
+    'hex'
```
it would be useful to see how this payload was created; or what the contents of it are. why not use `let` like we did above?

rafiss commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -6384,6 +6384,9 @@ table's zone configuration this will return NULL.`,
 					)
 				}
+				if err := evalCtx.Planner.CheckCanBecomeUser(evalCtx.Context, sd.User()); err != nil {
+					return nil, errors.Wrap(err, "cannot SET ROLE to user")
```
might be fine to return this error without extra wrapping. (the SET ROLE might be misleading anyway)

rafiss reviewed (approved):
lgtm, but requested some comments in the test

rafiss commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -6384,6 +6384,9 @@ table's zone configuration this will return NULL.`,
 						"can only deserialize matching session users",
 					)
 				}
+				if err := evalCtx.Planner.CheckCanBecomeUser(evalCtx.Context, sd.User()); err != nil {
```
actually, why don't we just make it so that `sd.CurrentUser().Normalized()` must match `evalCtx.SessionData().CurrentUser().Normalized()`?

it's a simpler check and we don't have any use case that needs to allow anyone to automatically change roles using this.

otan replied:
because during a session migration someone might use `SET ROLE`, e.g. using postgrest.

i don't think we should error that out

rafiss replied:
ah good call. i should also update `create_session_revival_token` so it uses the sessionUser rather than currentUser for the token then.

otan committed:
sql: check session_user can become user on deserialize_session

Release note (bug fix): `crdb_internal.deserialize_session` now checks
the session_user has the privilege to SET ROLE to the current_user before
changing the session settings.

otan commented:
bors r=rafiss
